xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/store/XWikiHibernateVersioningStore.java

```diff
@@ -147,11 +151,74 @@ private XWikiDocumentArchive getXWikiDocumentArchive(XWikiDocument doc, Revision
         XWikiContext inputxcontext) throws XWikiException
     {
         XWikiDocumentArchive archiveDoc = doc.getDocumentArchive();
-        // We only retrieve a cached archive if we want a complete one.
-        if (archiveDoc != null && criteria.isAllInclusive()) {
-            return archiveDoc;
+        if (archiveDoc == null) {
+            archiveDoc = getXWikiDocumentArchiveFromDatabase(doc, criteria, inputxcontext);
+        // if there's an archive doc and the criteria is to not return everything then we filter, else we just return
+        } else if (!criteria.isAllInclusive()) {
+            archiveDoc = filterArchiveFromCriteria(doc, archiveDoc, criteria);
         }
+        return archiveDoc;
+    }
+
+    private XWikiDocumentArchive filterArchiveFromCriteria(XWikiDocument doc, XWikiDocumentArchive archiveDoc,
+        RevisionCriteria criteria)
+    {
+        XWikiDocumentArchive result =
+            new XWikiDocumentArchive(doc.getDocumentReference().getWikiReference(), doc.getId());
+        Collection<XWikiRCSNodeInfo> nodes = archiveDoc.getNodes();
+        XWikiRCSNodeInfo nodeinfo = null;
+        List<XWikiRCSNodeInfo> results = new ArrayList<>();
+
+        // Iterate over all versions and get the ones matching the criteria
+        for (XWikiRCSNodeInfo nextNodeinfo : nodes) {
+            if (nodeinfo != null && (criteria.getIncludeMinorVersions() || !nextNodeinfo.isMinorEdit())) {
```

Comment:
So the original code was doing the same but I'm not sure to understand why we're checking `nextNodeInfo.isMinorEdit()` and not `nodeinfo.isMinorEdit()`: I don't see the point of getting 2 nodes here, instead of just doing a simple iteration over all nodes? @tmortagne do you remember why it was done like this?

Reply:
> @tmortagne do you remember why it was done like this?

Never really looked at this code before today to be honest.

Reply:
My bad I thought you contributed to it, was done by Sergiu in https://github.com/pjeanjean/xwiki-platform/commit/c1151265c9722102358e64a21d823bd234039d7f#diff-d9a61ffb6c31cdb34d3675069afabd6459d7a430a7de9c7917facbc9cd425df6 apparently. Not sure why though...

Reply:
I don't understand either why we check the minor edit status of the following node instead of the node we are actually adding in the filtered list. It feels like we will end up with the wrong nodes, but apparently it's not the case.

Reply:
OK, thanks to @pjeanjean finally understanding that the original code point was not to get the major nodes but the last nodes of each version "branch".